shiny/ui/_theme.py

```diff
@@ -0,0 +1,513 @@
+            f'@import "{path_mixins}";',
+            *self._mixins,
+            f'@import "{path_rules}";',
+            *self._rules,
+        ]
+
+        return "\n".join(sass_lines)
+
+    def to_css(
+        self,
+        compile_args: Optional[SassCompileArgs] = None,
+    ) -> str:
+        """
+        Compile the theme to CSS and return the result as a string.
+
+        Returns
+        -------
+        :
+            The compiled CSS for the theme. The value is cached such that previously
+            compiled themes are returned immediately. Adding additional custom Sass code
+            or changing the preset will invalidate the cache.
+        """
+        if self._css:
+            return copy.copy(self._css)
+
+        if self._can_use_precompiled():
+            self._css = self._read_precompiled_css()
+            return copy.copy(self._css)
+
+        check_libsass_installed()
+        import sass
+
+        if compile_args is None:
+            compile_args = {"output_style": "compressed"}
+
+        self._css = cast(
+            str,
+            sass.compile(
+                string=self.to_sass(),
+                include_paths=self._include_paths,
+                **compile_args,  # type: ignore
+            ),
+        )
+
+        return copy.copy(self._css)
+
+    def _read_precompiled_css(self) -> str:
+        path = path_pkg_preset(self._preset, "preset.min.css")
+        with open(path, "r") as f:
+            return f.read()
+
+    def _html_dependency_precompiled(self) -> HTMLDependency:
+        return HTMLDependency(
+            name=f"shiny-theme-{self._preset}",
+            version=self._version,
+            source={
+                "package": "shiny",
+                "subdir": f"www/shared/sass/preset/{self._preset}",
+            },
+            stylesheet={"href": "preset.min.css"},
+            all_files=False,
+        )
+
+    def _html_dependency(self) -> HTMLDependency:
+        """
+        Create an `HTMLDependency` object from the theme.
+
+        Returns
+        -------
+        :
+            An :class:`~htmltools.HTMLDependency` object representing the theme. In
+            most cases, you should not need to call this method directly. Instead, pass
+            the `Theme` object directly to the `theme` argument of any Shiny page
+            function.
+        """
+        if self._can_use_precompiled():
+            return self._html_dependency_precompiled()
+
+        dep_name = f"shiny-theme-{self.name or self._preset}"
+        css_name = f"{dep_name}.min.css"
+
+        # Re-use already compiled CSS file if possible
+        if self._css and self._css_temp_srcdir is not None:
+            return HTMLDependency(
+                name=dep_name,
+                version=Version(self._version),
+                source={"subdir": self._css_temp_srcdir},
+                stylesheet={"href": css_name},
+            )
+
+        tmpdir = tempfile.mkdtemp()
```

Comment:
I think we want to use `tempfile.TemporaryDirectory()` for this since that'll get removed automatically when Python exits

Reply:
Done in 7ec22755303e96bfe9b0b616d49af2245cac7548. I also simplified the logic so we track the `TemporaryDirectory` directly and we call its `.cleanup()` method whenever we generate a new copy.